### DISCUSSION TITLE
How to implement nested shortcut arguments?

### DISCUSSION THREAD

**Discussion by schulle4u** [datenstrom/community#447] [Ask a question] [answered]
Hello, 

is it possible to have nested arguments for shortcuts? I'm currently updating one of my extensions and would like to have some kind of subarguments for certain arguments in the shortcut. The shortcut calls javascript widgets, and a recently added widget has more arguments than others. Here is a code excerpt: 

```php
// Handle page content of shortcut
    public function onParseContentShortcut($page, $name, $text, $type) {
        $output = NULL;
        if ($name=="radioboss" && ($type=="block" || $type=="inline")) {
            list($type, $style, $server, $port, $port2, $id, $mount) = $this->yellow->toolbox->getTextArgs($text);
            // [...] usual handling of empty values and internal arguments

            $output .= "<div class=\"".htmlspecialchars($style)."\">\n";
            
            // Widget code
            if ($type == "miniplayer") {
                list($width, $activecolor, $greycolor) = $this->yellow->toolbox->getTextArgs($text);
                if (strempty($width)) $width = "82";
                if (empty($activecolor)) $activecolor = "#111111";
                if (empty($greycolor)) $greycolor = "#b3b3b3";
                $output .= "<!-- RadioBOSS Cloud Player Widget (Start) -->\n";
                $output .= "<div id=\"rbcloud_mplayer".htmlspecialchars($wid)."\"></div>\n";
                $output .= "<script src=\"https://".htmlspecialchars($server)."/w/mplayer.js?u=https%3A%2F%2F".htmlspecialchars($server)."%3A".htmlspecialchars($port2)."%2F".htmlspecialchars($mount)."&amp;wid=".htmlspecialchars($wid)."&amp;pw=".htmlspecialchars($width)."&amp;ca=".rawurlencode($activecolor)."&amp;cg=".rawurlencode($greycolor)."\"></script>\n";
                $output .= "<!-- RadioBOSS Cloud Player Widget (End) -->\n";
            }
    // [...] more argument handling
```

If needed, [here is the full code](http://github.com/schulle4u/yellow-extensions-schulle4u/tree/radioboss-dev/radioboss/radioboss.php). The shortcut partly works, but ignores the first argument from the subarguments list. What am I doing wrong here? The image extension uses similar methods to call certain functions, though this isn't exactly what I am looking for. I surely could give every widget it's own shortcut, but this would end up in to many code. 

Thanks in advance. 
Steffen

**Comment by markseuffert** [comment]
Do you want to handle a variable number of arguments? This is possible. You can parse the shortcut text a second time and handle more/different arguments. Here are examples:

```php
// Parse normal arguments
list($type, $style, $server, $port, $port2, $id, $mount) = $this->yellow->toolbox->getTextArgs($text);

// Parse more arguments
list($type, $style, $server, $port, $port2, $id, $mount, $width, $activecolor, $greycolor) = $this->yellow->toolbox->getTextArgs($text);

// Parse different arguments
list($type, $width, $activecolor, $greycolor) = $this->yellow->toolbox->getTextArgs($text);
```

Does this solve your problem?

**Comment by schulle4u** [comment]
Hmm, not sure if this solves the problem. I can of course append more arguments to the end, but wouldn't this make the shortcut more difficult to write for users? 
Ideally the shortcut should look like this: 

Normal shortcut: `[radioboss player $style $server...]`

Extra arguments for the miniplayer widget: `[radioboss miniplayer (86 #111111 #b3b3b3) $style $server...]`

The extra brackets are just for illustration. Maybe I'm thinking to complicated here again. 😀

**Comment by GiovanniSalmeri** [comment]
Perhaps you could do something like this?

`[radioboss "miniplayer 86 #111111 #b3b3b3" $style $server...]`

With the quotes, `miniplayer 86 #111111 #b3b3b3` counts as a single argument. Then you can trivially parse it in order to separate the subarguments.

**Comment by schulle4u** [comment]
@GiovanniSalmeri Thanks for your suggestion, this looks interesting. Maybe I should try to use `explode()` for this as a simple workaround.

**Comment by schulle4u** [comment, marked as answer]
Problem solved, thanks guys. 

```php
if ($widget == "miniplayer") {
    if (empty($arguments)) $arguments = "86 #111111 #b3b3b3";
    $argumentData = explode(" ", $arguments);
    $output .= "<!-- RadioBOSS Cloud Player Widget (Start) -->\n";
    $output .= "<div id=\"rbcloud_mplayer".htmlspecialchars($wid)."\"></div>\n";
    $output .= "<script src=\"https://".htmlspecialchars($server)."/w/mplayer.js?u=https%3A%2F%2F".htmlspecialchars($server)."%3A".htmlspecialchars($port2)."%2F".htmlspecialchars($mount)."&amp;wid=".htmlspecialchars($wid)."&amp;pw=".htmlspecialchars($argumentData[0])."&amp;ca=".rawurlencode($argumentData[1])."&amp;cg=".rawurlencode($argumentData[2])."\"></script>\n";
    $output .= "<!-- RadioBOSS Cloud Player Widget (End) -->\n";
}
```

Works for the moment, further testing tomorrow. 

Steffen
